### ISSUE TITLE
unit flake: TestEventf is racy

### ISSUE THREAD

**Issue by wojtek-t** [kubernetes/kubernetes#23533]
WARNING: DATA RACE
Write by goroutine 32:
  testing.(*common).log()
      /usr/local/go/src/testing/testing.go:353 +0x1e6
  testing.(*common).Logf()
      /usr/local/go/src/testing/testing.go:366 +0x82
  testing.(*common).Logf-fm()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/record/event_test.go:351 +0x5f
  k8s.io/kubernetes/pkg/client/record.(*eventBroadcasterImpl).StartLogging.func1()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/record/event.go:216 +0x2a7
  k8s.io/kubernetes/pkg/client/record.(*eventBroadcasterImpl).StartEventWatcher.func1()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/record/event.go:237 +0x14e

Previous read by main goroutine:
  runtime.convT2E()
      /usr/local/go/src/runtime/iface.go:128 +0x0
  testing.(*T).report()
      /usr/local/go/src/testing/testing.go:538 +0xa18
  testing.RunTests()
      /usr/local/go/src/testing/testing.go:591 +0xbdf
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:515 +0x11d
  main.main()
      k8s.io/kubernetes/pkg/client/record/_test/_testmain.go:70 +0x210

Goroutine 32 (running) created at:
  k8s.io/kubernetes/pkg/client/record.(*eventBroadcasterImpl).StartEventWatcher()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/record/event.go:239 +0x8f
  k8s.io/kubernetes/pkg/client/record.(*eventBroadcasterImpl).StartLogging()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/record/event.go:217 +0xa5
  k8s.io/kubernetes/pkg/client/record.TestEventf()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/record/event_test.go:351 +0x3034
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:473 +0xdc

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/23527/kubernetes-pull-test-unit-integration/19502/build-log.txt

**Comment by ncdc** [comment]
I think this is an issue with go 1.6. `testing.(T).report()` doesn't grab the lock for `output` like it should so it doesn't race with `testing.(common).log()`. It looks like this has maybe been fixed in go by https://go-review.googlesource.com/#/c/18893/.

**Comment by a-robinson** [comment]
And again:
https://cloud.google.com/console/storage/kubernetes-jenkins/logs/kubernetes-test-go/10362/

```
	event_test.go:351: Event(api.ObjectReference{Kind:"Pod", Namespace:"baz", Name:"foo", UID:"bar", APIVersion:"version", ResourceVersion:"", FieldPath:"spec.containers[2]"}): type: 'Normal' reason: 'Started' some verbose message: 1
	events_cache_test.go:97: � - expectedEvent.Count is 1
	event_test.go:351: Event(api.ObjectReference{Kind:"Pod", Namespace:"baz", Name:"foo", UID:"bar", APIVersion:"version", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'Killed' some other verbose message: 1
	events_cache_test.go:97:  - expectedEvent.Count is 1
	event_test.go:351: Event(api.ObjectReference{Kind:"Pod", Namespace:"baz", Name:"foo", UID:"bar", APIVersion:"version", ResourceVersion:"", FieldPath:"spec.containers[2]"}): type: 'Normal' reason: 'Started' some verbose message: 1
	events_cache_test.go:97:  - expectedEvent.Count is 2
	event_test.go:351: Event(api.ObjectReference{Kind:"Pod", Namespace:"baz", Name:"foo", UID:"differentUid", APIVersion:"version", ResourceVersion:"", FieldPath:"spec.containers[3]"}): type: 'Normal' reason: 'Started' some verbose message: 1
	events_cache_test.go:97:  - expectedEvent.Count is 1
	event_test.go:351: Event(api.ObjectReference{Kind:"Pod", Namespace:"baz", Name:"foo", UID:"bar", APIVersion:"version", ResourceVersion:"", FieldPath:"spec.containers[2]"}): type: 'Normal' reason: 'Started' some verbose message: 1
	events_cache_test.go:97:  - expectedEvent.Count is 3
	event_test.go:351: Event(api.ObjectReference{Kind:"Pod", Namespace:"baz", Name:"foo", UID:"differentUid", APIVersion:"version", ResourceVersion:"", FieldPath:"spec.containers[3]"}): type: 'Normal' reason: 'Stopped' some verbose message: 1
	events_cache_test.go:97:  - expectedEvent.Count is 1
	events_cache_test.go:97:  - expectedEvent.Count is 2
==================
WARNING: DATA RACE
Write by goroutine 35:
  testing.(*common).log()
      /usr/local/go/src/testing/testing.go:353 +0x1e6
  testing.(*common).Logf()
      /usr/local/go/src/testing/testing.go:366 +0x82
  testing.(*common).Logf-fm()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/record/event_test.go:351 +0x5f
  k8s.io/kubernetes/pkg/client/record.(*eventBroadcasterImpl).StartLogging.func1()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/record/event.go:216 +0x2a7
  k8s.io/kubernetes/pkg/client/record.(*eventBroadcasterImpl).StartEventWatcher.func1()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/record/event.go:237 +0x14e

Previous read by main goroutine:
  runtime.convT2E()
      /usr/local/go/src/runtime/iface.go:128 +0x0
  testing.(*T).report()
      /usr/local/go/src/testing/testing.go:538 +0xa18
  testing.RunTests()
      /usr/local/go/src/testing/testing.go:591 +0xbdf
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:515 +0x11d
  main.main()
      k8s.io/kubernetes/pkg/client/record/_test/_testmain.go:70 +0x210

Goroutine 35 (running) created at:
  k8s.io/kubernetes/pkg/client/record.(*eventBroadcasterImpl).StartEventWatcher()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/record/event.go:239 +0x8f
  k8s.io/kubernetes/pkg/client/record.(*eventBroadcasterImpl).StartLogging()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/record/event.go:217 +0xa5
  k8s.io/kubernetes/pkg/client/record.TestEventf()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/record/event_test.go:351 +0x3034
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:473 +0xdc
```

**Comment by wojtek-t** [comment]
This is happening pretty often now.

@kubernetes/sig-api-machinery 
@caesarxuchao @lavalamp - can one of you maybe take a look? Maybe we can comment this test out for some time?

**Comment by ncdc** [comment]
I don't think this can be fixed until go 1.6.1 (assuming that contains the patch I linked above). The only other option I see is to disable the test for now.

**Comment by ncdc** [comment]
In case it's not evident, the unit tests are running with go 1.6.

**Comment by caesarxuchao** [comment]
@ncdc it looks like this bug should affect all tests that call t.Logf?

**Comment by caesarxuchao** [comment]
I think we can just to comment out the line that calls t.Logf https://github.com/kubernetes/kubernetes/blob/master/pkg/client/record/event_test.go#L351 and it won't affect the test logic.

**Comment by ncdc** [comment]
The issue with this particular test case is that a goroutine is trying to log at the same time the testing framework is trying to record pass/fail. So I think anything that is logging in a goroutine at the same time pass/fail is being recorded.

**Comment by ncdc** [comment]
@caesarxuchao I agree that commenting out that line would fix the flake and it won't affect the test logic. I think we have this pattern in at least 1 other test too (maybe it's the HPA one that's failing, I can't remember).

**Comment by wojtek-t** [comment]
Yeah - what @caesarxuchao suggested looks fine
@caesarxuchao - can you send out a PR?

**Comment by caesarxuchao** [comment]
Sure. I'll send a fix for event_test.go. Thank you both for the confirmation.

@ncdc I couldn't find an issue for HPA. We can fix that when it's reported.

**Comment by ncdc** [comment]
@caesarxuchao I was thinking of #23361 but that might be different.

**Comment by wojtek-t** [comment]
From the stacktrace - it seems different.

**Comment by ncdc** [comment]
Looks like we hit this again in https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/29457/kubernetes-pull-test-unit-integration/36626/

**Comment by rootfs** [comment]
looks another hit here https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/24977/kubernetes-pull-test-unit-integration/37130/

**Comment by caesarxuchao** [comment]
Fixed by #30033.
